src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.InterpolatedStringContext.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.CodeAnalysis.FlowAnalysis
+{
+    internal partial class ControlFlowGraphBuilder
+    {
+        private readonly struct InterpolatedStringHandlerContext
```

Comment:
>InterpolatedStringHandlerContext [](http://example.com/codeflow?start=32&length=32)

It feels like we have a certain amount of over-engineering here. In particular:
- I don’t think we would want to inline this structure into ```ControlFlowGraphBuilder```. A class would work just fine, especially given that ```IInterpolatedStringHandlerCreationOperation```s won’t be common.
- The implementation needs to worry about default values and has to explicitly track the state to recognize them.
- Flip-flopping between “single” and “multiple” mode adds a fair amount of complexity to the implementation and makes it harder to keep track of interaction between different parts of the builder. I haven’t finished reviewing all changes, but it looks like the only advantage we get from the “single” mode is the ability to assert the ``` IInterpolatedStringHandlerCreationOperation``` node being visited. So, it is not instrumental to the algorithm overall. In my opinion the added complexity does not outweigh the benefit to be able to do ```Debug.Assert```. The same can be achieved by adding a DEBUG-only field to the context, that is set/cleared in ```VisitInterpolatedStringHandlerCreation```.
- Flip-flopping between the ```IInterpolatedStringHandlerCreationOperation``` and ```IArgumentOperation``` also adds complexity.

Bottom line, I think we should switch to class to represent the context. Some portion of the, context that is initialized when we preparing to visit arguments, is read-only. The other part of the context is mutable and specific to particular ```IInterpolatedStringHandlerCreationOperation``` node we are visiting. That portion is set/cleared in ```VisitInterpolatedStringHandlerCreation```. The parts can be split in two different classes and can be stored in two different fields of ```ControlFlowGraphBuilder```, but I do not see any big advantage of doing this. All context that ```ControlFlowGraphBuilder``` is mutable. I don’t see much difference between modifying fields directly stored in ```ControlFlowGraphBuilder``` and modifying fields stored in classes stored in fields of ```ControlFlowGraphBuilder```. So, going after read-only fields is a non-goal when we are talking about the context.

Reply:
I can try to simplify the implementation by splitting it up into two classes, but I think a partially-mutable class where some of the fields apply to a specific set of interpolated string creations and some of the fields apply to the _current_ creation (regardless of whether the current creation is in that previous list) is more confusing than the current implementation.